src/lib.rs

```diff
@@ -27,102 +28,1070 @@ pub extern "C" fn wapc_init() {
     register_function("protocol_version", protocol_version_guest);
 }
 
+fn is_valid_id(id: i64, ranges: &[IDRange]) -> bool {
+    for range in ranges {
+        if id >= range.min && id <= range.max {
+            return true;
+        }
+    }
+    false
+}
```

Comment:
I would move that to make it a public method of `RuleStrategy`. It feels cleaner to me